### ISSUE TITLE
Can't drag Draggable over a WebViewWidget

### ISSUE THREAD

**Issue by M1loseph** [flutter/flutter#142427]
### Steps to reproduce

1. Install packages `fluttter_webview` and `flutter_webview_web`.
2. Run the provided code sample on chrome.
3. Try to drag the red square over the white space - it should work.
4. Drag the square over WeView - this time around square stops on the edge of WebView and doesn't follow the cursor.

### Expected results

Draggable can be dragged over WebView that is below in the Stack widget.

### Actual results

Draggable stops on the edge of WebView and doesn't follow the mouse cursor. 

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

// #1
void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final WebViewController _controller = WebViewController()
    ..loadRequest(
      Uri.parse('https://flutter.dev'),
    );
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        useMaterial3: true,
      ),
      home: Scaffold(
        body: Stack(
          children: [
            Center(
              child: ConstrainedBox(
                constraints: const BoxConstraints(
                  maxHeight: 500,
                  maxWidth: 500,
                ),
                child: WebViewWidget(
                  controller: _controller,
                ),
              ),
            ),
            Draggable(
              child: Container(
                width: 100,
                height: 100,
                color: Colors.red,
              ),
              feedback: Container(
                width: 100,
                height: 100,
                color: Colors.red,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

![image](https://github.com/flutter/flutter/assets/53463658/93867179-3ca0-4b67-899b-7e9015611e8f)

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.16.7, on Microsoft Windows [Version 10.0.19045.3930], locale en-GB)
    • Flutter version 3.16.7 on channel stable at C:\Users\user\dev\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ef1af02aea (2 weeks ago), 2024-01-11 15:19:26 -0600
    • Engine revision 4a585b7929
    • Dart version 3.2.4
    • DevTools version 2.28.5

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)      
    • Android SDK at **anonymized**
    • Platform android-33, build-tools 34.0.0
    • ANDROID_HOME = **anonymized**
    • ANDROID_SDK_ROOT = **anonymized**
    • Java binary at: **anonymized**
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-b2043.56-10550314)     
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at **anonymized**

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.8.4)        
    • Visual Studio at **anonymized**
    • Visual Studio Community 2022 version 17.8.34408.163
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2023.1)
    • Android Studio at **anonymized**
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-b2043.56-10550314)     

[√] IntelliJ IDEA Community Edition (version 2023.3)
    • IntelliJ at **anonymized**
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] VS Code (version 1.85.2)
    • VS Code at **anonymized**
    • Flutter extension version 3.80.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.3930]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 121.0.6167.85
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 121.0.2277.83

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by M1loseph** [comment]
I tried to wrap `Draggable` with `PointerInterceptor` from `pointer_interceptor` package, but it didn't help.

**Comment by huycozy** [comment]
Thanks for filing the issue, @M1loseph 
Looks like you typed the wrong package name, it should be [webview_flutter](https://pub.dev/packages/webview_flutter) and [webview_flutter_web](https://pub.dev/packages/webview_flutter_web) :)

Anyway, I can reproduce this issue on Web.

https://github.com/flutter/flutter/assets/104349824/ca76ec1d-452f-4b73-b33d-744ddbee090e

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.16.9, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.16.9 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 41456452f2 (14 hours ago), 2024-01-25 10:06:23 -0800
    • Engine revision f40e976bed
    • Dart version 3.2.6
    • DevTools version 2.28.5

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio Hedgehog.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.85.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D                         • android-arm64  • Android 11 (API 30)
    • iPhone (mobile)  • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 15.8 19H370
    • macOS (desktop)  • macos                                    • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome                                   • web-javascript • Google Chrome 120.0.6099.234

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.19.0-9.0.pre.204, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.19.0-9.0.pre.204 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efb1346767 (32 hours ago), 2024-01-27 14:07:24 -0500
    • Engine revision f3d48be766
    • Dart version 3.4.0 (build 3.4.0-81.0.dev)
    • DevTools version 2.31.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio Hedgehog.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.85.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 121.0.6167.85

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by M1loseph** [comment]
Yes, you're right. Thank you for noticing misspelled packages. I updated the description 😄.

**Comment by wanjm** [comment]
I think it is that when mouse on iframe, the event is send to iframe instead of current widget;
i we add button on top of webview widget, the button is not clickable;

**Comment by ditman** [comment]
What's the expected behavior of this on "drop"?

One way I can think of doing this is putting a PointerInterceptor *on top* of the HtmlElementView while dragging the Draggable, and remove it after dragging is done?

(Dragging the PointerInterceptor is not a bad idea either, but I'm not sure how those two widgets would interact together!)

**Comment by wanjm** [comment]
1.  I wrap PointerInterceptor on WebViewWidget, it doesn't work;
2. I warp WebViewWidget on FilledButton in my example, but button is clickable;

and I found that PointerInterceptor create a empty platformview on top of its child;  then I change the code of PointerInterceptor to put the empty platformview under it child, then case one works;  @M1loseph 
```
    return Stack(
      alignment: Alignment.center,
      children: <Widget>[
        child,  /// this line used to after Position;
        Positioned.fill(
          child: HtmlElementView(
            viewType: viewType,
          ),
        ),
      ],
    );
```

could you please let me know why HtmlElementView should be put on top of child, can it be put under child;  what is the mouse event transmit order? I just try to understand why does it work;

1. can the PointerInterceptor be merged with WebView_flutter as we are now using it;

**Comment by wanjm** [comment]
![WeChatWorkScreenshot_b71ef27c-f41f-4b54-a1fd-d464de89eb61](https://github.com/flutter/flutter/assets/3118573/d9aea0fa-0fe2-46fb-980f-4f0b55020da7)
I added line 82 also make it work;

**Comment by wanjm** [comment]
@ditman  as PointerInterceptor tells, iOS also has such problem, but when i test the case using iPad, it works, not issue occurred; 
Is there any example that  PointerInterceptor is needed in iOS;

**Comment by ditman** [comment]
@wanjm the explanation of how PointerInterceptor works is in its README:

* https://pub.dev/packages/pointer_interceptor

(It works by placing an empty HTML widget **behind its child** ([code](https://github.com/flutter/packages/blob/main/packages/pointer_interceptor/pointer_interceptor_web/lib/pointer_interceptor_web.dart#L37-L48)). (There's a request to flip the order and [put the empty html element on top](https://github.com/flutter/flutter/issues/121272)).

--- 

> can the PointerInterceptor be merged with WebView_flutter as we are now using it;

Hm, this is a bigger problem than the WebView plugin, everything in the framework needs the pointer interceptor. See https://github.com/flutter/flutter/issues/93887. (Needs help from the framework).

Adding `pointerEvents = 'none'` to the webview could be a good feature request? I'm not sure how that could be implemented on other platforms (it's basically preventing any mouse/pointer interaction with the view from doing anything)

---

> as there are two platform view now, does it have performance impact?

On the web there's a limit of platform views in some browsers (firefox, safari), but we're working to remove that limit / optimize performance. See https://github.com/flutter/flutter/issues/137953.

---

> when i test the case using iPad, it works, not issue occurred

I can't exactly remember what problem motivated the implementation of the plugin for iOS, but it might have been a different type of interaction with platform views, not necessarily the same use case as in the web!

---

TL;DR: it seems the trick in your case was making `pointerEvents='none'` in the iframe, @wanjm?

**Comment by wanjm** [comment]
> TL;DR: it seems the trick in your case was making `pointerEvents='none'` in the iframe, @wanjm?
yes, I add the line in my case, as in my case we don't need to interact with webview;

but I hope WebViewControllor should have one switch to control whether  WebView_flutter is interactable or not.  
and  when the WebView_flutter is not interactive, it can be optimized by engine in my opinion;  and the optimize can be done for all platform besides web; that is the reason I suggest WebViewController should have such switch;

**Comment by ditman** [comment]
> that is the reason I suggest WebViewController should have such switch

Please create a new issue to request this feature! https://github.com/flutter/flutter/issues/new

**Comment by M1loseph** [comment]
Isn't it a weird solution to this problem? 
As a user I expect that the WebView widget will behave kinda like a regular widget. If I dragged some widget over a button, the button would not be clicked all of the sudden. Is there something that I'm missing that makes it impossible to implement the same behaviour for WebView on the framework side?

**Comment by ditman** [comment]
> As a user I expect that the WebView widget will behave kinda like a regular widget.

The expectation is fair.

> Is there something that I'm missing that makes it impossible to implement the same behaviour for WebView on the framework side?

I currently don't have time to dive deep into this particular use case, so I was just suggesting solutions with things I already know exist, that might help you work around this issue without having to wait for a fix to the engine/web framework.

**Comment by wanjm** [comment]
@ditman 
what does "isVisible: false," in   ui_web.platformViewRegistry.registerViewFactory used for. 
as you tell in PointerInterceptor 0.9.2, what optimize will engine do.

**Comment by ditman** [comment]
@wanjm "visible" platform views cause flutter to finish drawing in a canvas, and start drawing in a new one (that is rendered on top of the visible platform view). In the DOM:

```
<flutter-view>
  <canvas>  <- potentially lots of draw commands
  <platform-view (visible)>  <- we need to render a platform view
  <canvas>  <- new canvas to continue drawing "on top" of the platform view
</flutter-view>
```

When the platform view is set to "invisible", we allow multiple "invisible" platform views to be rendered without breaking the canvas, so something like:

```
<flutter-view>
  <canvas>
  <platform-view (invisible)>
  <platform-view (invisible)>
  <platform-view (visible)>
  <canvas>
</flutter-view>
```

In some browsers, the number of "canvases" (actually GL Contexts) is limited, so that optimization lets flutter use more "invisible" platform views (like those in the pointer interceptor, or the link widget) without splitting the drawing into too many contexts.

This test may explain it better than I do:

* https://github.com/flutter/engine/blob/main/lib/web_ui/test/canvaskit/embedded_views_test.dart#L772

**Comment by wanjm** [comment]
@ditman thanks for you explanation;
if there are many platform-view visible, the dom tree may look like following, right?
```
<flutter-view>
  <canvas>
  <platform-view (visible)>
<canvas>
  <platform-view (visible)>
  <canvas>
  <platform-view (visible)>
  <canvas>
</flutter-view>
```

**Comment by wanjm** [comment]
@ditman why it is said that PlatformView is expensive; does it expensive on web; for i found that is implemented using slot, I thought it is not expensive;

**Comment by ditman** [comment]
> if there are many platform-view visible, the dom tree may look like following, right?

@wanjm Correct (check the other unit tests!). There's a limit of 7 "overlays", because for some (non-chromium) browsers, every time we change the canvas we need a new WebGL Context.

> it is said that PlatformView is expensive; does it expensive on web; for i found that is implemented using slot, I thought it is not expensive

No, on the web platform, they aren't too expensive, but it depends on what you put into the platform view (same as with any other web app). In general, `iframe` is the most expensive thing to put in a platform view ([see the MDN warning](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)), but anything that is "heavy" in another web app, will also be heavy in your Flutter Web app.

**Comment by Gustl22** [comment]
The main problem here is: the platform view is integrated via an absolute positioned window/div, which can't passthrough only a selection of pointer events, at least I'm not aware of a method:
https://stackoverflow.com/questions/1009753/pass-mouse-events-through-absolutely-positioned-element
So we need some mechanism to pass these events without loosing interativity with the webview.
Also making the iframe transparent did not really help: https://stackoverflow.com/questions/3064199/transparent-iframe-body
